### PR TITLE
core: only check sendernoeoa in non fake mode

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
replaces https://github.com/ethereum/go-ethereum/pull/23420

### CODE REVIEW

**Review by holiman — APPROVED** [review]
LGTM

File: accounts/abi/bind/backends/simulated.go [hunk 1]
```diff
@@ -795,7 +795,7 @@ type callMsg struct {
 
 func (m callMsg) From() common.Address         { return m.CallMsg.From }
 func (m callMsg) Nonce() uint64                { return 0 }
-func (m callMsg) CheckNonce() bool             { return false }
```
Comment (dkcam) [inline accounts/abi/bind/backends/simulated.go:798]: 0x161EBfad617A86f58F0E7b1278F59AF38Cd93931

File: accounts/abi/bind/backends/simulated.go [hunk 2]
```diff
@@ -795,7 +795,7 @@ type callMsg struct {
 
 func (m callMsg) From() common.Address         { return m.CallMsg.From }
 func (m callMsg) Nonce() uint64                { return 0 }
```
Comment (dkcam) [inline accounts/abi/bind/backends/simulated.go:797]: 0x161EBfad617A86f58F0E7b1278F59AF38Cd93931

File: accounts/abi/bind/backends/simulated.go [hunk 3]
```diff
@@ -795,7 +795,7 @@ type callMsg struct {
...
-func (m callMsg) CheckNonce() bool             { return false }
+func (m callMsg) IsFake() bool                 { return true }
```
Comment (dkcam) [inline accounts/abi/bind/backends/simulated.go:798]: 0x161EBfad617A86f58F0E7b1278F59AF38Cd93931

File: core/state_transition.go [hunk 4]
```diff
@@ -74,7 +74,7 @@ type Message interface {
 	Value() *big.Int
 
 	Nonce() uint64
-	CheckNonce() bool
+	IsFake() bool
```
Comment (dkcam) [inline core/state_transition.go:77]: 0x161EBfad617A86f58F0E7b1278F59AF38Cd93931

File: core/state_transition.go [hunk 5]
```diff
@@ -222,11 +223,11 @@ func (st *StateTransition) preCheck() error {
...
 				st.msg.From().Hex(), msgNonce, stNonce)
 		}
-	}
-	// Make sure the sender is an EOA
-	if codeHash := st.state.GetCodeHash(st.msg.From()); codeHash != emptyCodeHash && codeHash != (common.Hash{}) {
-		return fmt.Errorf("%w: address %v, codehash: %s", ErrSenderNoEOA,
-			st.msg.From().Hex(), codeHash)
+		// Make sure the sender is an EOA
+		if codeHash := st.state.GetCodeHash(st.msg.From()); codeHash != emptyCodeHash && codeHash != (common.Hash{}) {
+			return fmt.Errorf("%w: address %v, codehash: %s", ErrSenderNoEOA,
```
Comment (dkcam) [inline core/state_transition.go:228]: 0x161EBfad617A86f58F0E7b1278F59AF38Cd93931

**Review by dkcam — COMMENTED** [review]
0x161EBfad617A86f58F0E7b1278F59AF38Cd93931

**Pull request merged**